eZ/Bundle/EzPublishCoreBundle/Entity/EntityManagerFactory.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Bundle\EzPublishCoreBundle\Entity;
+
+use Doctrine\ORM\EntityManagerInterface;
+use eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryConfigurationProvider;
+use Symfony\Component\DependencyInjection\ServiceLocator;
+
+/**
+ * @internal
+ */
+class EntityManagerFactory
+{
+    /** @var \eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryConfigurationProvider */
+    private $repositoryConfigurationProvider;
+
+    /** @var \Symfony\Component\DependencyInjection\ServiceLocator */
+    private $serviceLocator;
```

Comment:
I actually don't see any particular advantage of using `ContainerInterface` here. Especially when we have to use PSR11 `ContainerInterface`. From what I can see we use the Interface from Symfony DependencyInjection component in other places. Also `ServiceLocator` shows cleaner intention that you are only going to fetch some services from it. It would be confusing to some people that we basically got rid of the Container use but we somehow are still using it directly. :-1: if you ask me.

Reply:
It's just that `ServiceLocator` extends `ContainerInterface`. We don't use the functionalities specific to `ServiceLocator` itself (which is literally one method for listing provided services). Since `ContainerInterface` brings the least required abstraction for the `EntityManagerFactory`, that's why I suggested type hinting against just it, not the extending implementaton. Isn't it what should be done? Type hint against the most abstract interface that works?

At the very least use `\Symfony\Contracts\Service\ServiceProviderInterface`, which the ServiceLocator is actually an implementation of.

Associating `ContainerInterface` instance with main application container only is simply incorrect. Symfony default's `AbstractController` is type-hinted to contain `ContainerInterface`, which in reality is a `ServiceLocator`. The same goes for `\Symfony\Contracts\Service\ServiceSubscriberTrait`.

Reply:
I agree with @webhdx that it might be confusing, especially to developers who started their adventure with Symfony v2. I wouldn't also treat everything which Symfony does as dogma. There's a lot of inconsistency and contradiction in recent years of the framework.

Reply:
`\Symfony\Contracts\Service\ServiceProviderInterface` is not available in Symfony 3.4. I will try to incorporate this change when merging up.